editorial review of recent changes to migrating to native delivery (#5219)
Particular/docs.particular.net: nservicebus/tools/migrate-to-native-delivery.md

- [RavenDB persistence](/persistence/ravendb) versions 3.5.x and 4.x of the RavenDB database server
- [Azure Storage persistence](/persistence/azure-table) version 2.4.x compatible timeouts stored in Azure Storage tables

## Supported transports

The tool supports the following transports:

- [RabbitMQ transport](/transports/rabbitmq/)
- [SQL transport](/transports/sql)
- [Azure Storage Queues transport](/transports/azure-storage-queues/)

> 
> NOTE: Support for the [Amazon SQS transport](/transports/sqs/) will be added before NServiceBus version 8 is released.

Reviewer: ```suggestion
```

Reviewer: This can be removed completely as we don't need to support Amazon SQS in the tool.

- `--databaseName`: The database name where timeouts to migrate are stored
- `--ravenVersion`: Allowed values are "3.5" and "4"
- `--prefix`(optional): The prefix used for storage of timeouts. The default value is "TimeoutDatas"

> - `--forceUseIndex`(Optional): Required when migrating large amounts or timeouts. Requires all endpoints using the database to be turned off so as not to modify the timeout data. This option will only be used during migration.

Author: I didn't verify this, but it seems to follow?

> - `--forceUseIndex`(Optional): Required when migrating large amounts or timeouts. Requires all endpoints using the database to be turned off so as not to modify the timeout data. This option will only be used during migration.

Reviewer: ```suggestion
- `--forceUseIndex`(Optional): Required when migrating large amounts of timeouts. Requires all endpoints using the database to be turned off so as not to modify the timeout data. This option will only be used during migration.
```

### NHibernate persistence

>  - Delete the `StagedTimeoutEntity` table. This table contains copies of migrated timeouts.

Author: @lailabougria @danielmarbach I went ahead and changed this line to instruct the user to delete the entire table. Is that correct? I wasn't entirely sure from the previous conversation thread.

Reviewer: Yes, as part of the cleanup after the migration, it does make sense to delete it.

### NHibernate persistence

 - Delete all rows from the `StagedTimeoutEntity` table. These are copies of migrated timeouts.

>  - Delete the `MigrationsEntity` table. Each row in this table represents a previously performed migration.

Author: Can the entire table be deleted or just the rows?

Reviewer: The entire table may be deleted once the migration is complete. It's a one-time thing, we have it in place only for troubleshooting or verification purposes.

Author: Sorry, I put this comment on the wrong line. I was referring to the `StagedTimeoutEntity` table. Can that table be deleted or just the rows?

### Azure Storage persistence

When all the timeouts in a given timeout table have been migrated, the timeout table may be deleted. It is also possible to delete only migrated entities from a timeout table by deleting all rows with `OwningTimeoutManager` starting with `__hidden__`.

> For safety reasons, the migration table `timeoutsmigration` is left in the state of the last migration that was run. The table is cleaned automatically during every migration run. To reduce storage costs, after all migrations are done it is advisable to delete:

Author: > For safety reasons...

Can we be more specific? Is it for visibility?

Reviewer: I'm guessing this was aimed at tool-recovery/verification scenarios. If we would delete this table at the end of a successful migration, there's no way for the user to verify what was migrated in detail. @danielmarbach @WojcikMike Thoughts?

Author: 👍 thanks, maybe:

```suggestion
The `timeoutsmigration` table is left in the state of the last migration that was run in case it is required for troubleshooting. The table is cleaned automatically during every migration run. To reduce storage costs, after all migrations are done it is advisable to delete:
```

`SELECT * FROM TimeoutsMigration_State`

To list the status of timeouts for a previous/in-progress run, take the `MigrationRunId` from the results of the previous query and execute:

`SELECT * FROM TimeoutData_migration_{MigrationRunId}`

The results include the batch number and its status: `0=Pending`, `1=Staged`, or `2=Completed`.

### NHibernate persistence

The history and migrated data is always kept in the database.

To list the history and status of migrations:

`SELECT * FROM MigrationsEntity`

To list all the timeouts that were staged for migration:

`SELECT * FROM StagedTimeoutEntity` 

The results include the batch number and its status: `0=Pending`, `1=Staged`, or `2=Completed`.

### Azure Storage persistence

To list the history and status of migrations open the `timeoutsmigrationtoolstate` Azure Table inside the Storage Explorer.

The migrated timeouts and their status are contained in the `timeoutsmigration` Azure Table. This includes the batch number its status: `0=Pending`, `1=Staged`, or `2=Completed`.

> To list all timeouts that were transferred from the endpoints timeout table to the `timeoutsmigration` table, execute the following query and adjust the partition key cut-off time accordingly. To avoid returning additional timeout data, scope the partition key to the cutoff time plus one hundred years.

Author: > To avoid returning additional timeout data...

What is this "additional" timeout data? Can we be more descriptive? What are these rows that are more than 100 years out?

Reviewer 2: @adamralph the problem here is not timeouts that are more than 100 years old the problem is how the persisted works and how querying azure table including its weird lexicographical ordering operates. We could be more specific but if we do we end up explaining the internal legacy design of the persisted and need to explain query semantics of azure table which I find not helpful for the purpose of this document
